server/libs/storage/src/storage.service.ts

```diff
@@ -0,0 +1,110 @@
+import { APP_UPLOAD_LOCATION } from '@app/common';
+import { AssetEntity } from '@app/database/entities/asset.entity';
+import { ImmichConfigService } from '@app/immich-config';
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { FileSystemStorageService } from './storage.service.filesystem';
+import handlebar from 'handlebars';
+import * as luxon from 'luxon';
+import sanitize from 'sanitize-filename';
+import path from 'node:path';
+import { constants } from 'node:fs';
+import fsPromise from 'fs/promises';
+import mv from 'mv';
+
+@Injectable()
+export class StorageService {
+  readonly log = new Logger(StorageService.name);
+
+  constructor(
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+
+    private fileSystemStorageService: FileSystemStorageService,
+
+    private immichConfigService: ImmichConfigService,
+  ) {}
+
+  public async moveFile(asset: AssetEntity, filename: string): Promise<AssetEntity> {
+    try {
+      const parsedPath = await this.buildPath(asset, filename);
+      let duplicateCount = 1;
+      let qualifiedPath = path.join(APP_UPLOAD_LOCATION, asset.userId, parsedPath);
+
+      while (await this.checkFileExist(qualifiedPath)) {
+        const newPath = await this.buildPath(asset, filename + `_${duplicateCount}`);
+        qualifiedPath = path.join(APP_UPLOAD_LOCATION, asset.userId, newPath);
+        duplicateCount++;
+      }
+
+      const destination = await this.moveAndDeleteFile(asset.originalPath, qualifiedPath);
+
+      asset.originalPath = destination;
+      return await this.assetRepository.save(asset);
+    } catch (e) {
+      this.log.error(e);
+      return asset;
+    }
+  }
+
+  private moveAndDeleteFile(original: string, destination: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+      mv(original, destination, { mkdirp: true, clobber: false }, (err) => {
+        if (err) {
+          reject(err);
+        }
+
+        resolve(destination);
+      });
+    });
+  }
+
+  private async checkFileExist(path: string): Promise<boolean> {
+    try {
+      await fsPromise.access(path, constants.F_OK);
+      return true;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  private async buildPath(asset: AssetEntity, filename: string) {
+    // Build path from user config
+    // Get user config from database
+    const configs = await this.immichConfigService.getConfig();
```

Comment:
This is going to do a database read for every file uploaded. We should cache this when the storage service is initialized.

Reply:
What is the proposed method for invalidation when the config is changed?

Reply:
I think adding an observable to the config service would work well. Something like this:
```typescript
// config.service.ts
import { Subject } from 'rxjs';
...

public config$ = new Subject();
...
  this.config$.next(newConfig); // in update method (maybe change update to load the new config and return it + emit it)
...

//  storage service
this.immichConfigService.config$.subscribe((config) => {
  // update template and config cached values
});
```

Reply:
StorageService registers a listener on ConfigServer that is called whenever the config changes